File: patterns/Creational/AbstractFactory/Go/RealWorldExample/WebTemplate.go
package main

import (
	"fmt"
	"strings"
)

// TemplateFactory - The Abstract Factory interface declares creation methods for each distinct product type.
type TemplateFactory interface {
	CreateTitleTemplate() TitleTemplate
	CreatePageTemplate() PageTemplate
	GetRenderer() TemplateRenderer
}

// TwigTemplateFactory - Concrete Factory for creating Twig templates.
type TwigTemplateFactory struct{}

func (f *TwigTemplateFactory) CreateTitleTemplate() TitleTemplate {
	return &TwigTitleTemplate{}
}

func (f *TwigTemplateFactory) CreatePageTemplate() PageTemplate {
	return &TwigPageTemplate{BasePageTemplate: BasePageTemplate{titleTemplate: f.CreateTitleTemplate()}}
}

func (f *TwigTemplateFactory) GetRenderer() TemplateRenderer {
	return &TwigRenderer{}
}

// PHPTemplateFactory - Concrete Factory for creating PHP templates.
type PHPTemplateFactory struct{}

func (f *PHPTemplateFactory) CreateTitleTemplate() TitleTemplate {
	return &PHPTemplateTitleTemplate{}
}

func (f *PHPTemplateFactory) CreatePageTemplate() PageTemplate {
	return &PHPTemplatePageTemplate{BasePageTemplate: BasePageTemplate{titleTemplate: f.CreateTitleTemplate()}}
}

func (f *PHPTemplateFactory) GetRenderer() TemplateRenderer {
	return &PHPTemplateRenderer{}
}

// TitleTemplate - The Abstract Product interface for title templates.
type TitleTemplate interface {
	GetTemplateString() string
}

// TwigTitleTemplate - Concrete Product for Twig title templates.
type TwigTitleTemplate struct{}

func (t *TwigTitleTemplate) GetTemplateString() string {
	return "<h1>{{ title }}</h1>"
}

// PHPTemplateTitleTemplate - Concrete Product for PHP title templates.
type PHPTemplateTitleTemplate struct{}

func (t *PHPTemplateTitleTemplate) GetTemplateString() string {
	return "<h1><?= $title; ?></h1>"
}

// PageTemplate - The Abstract Product interface for page templates.
type PageTemplate interface {
	GetTemplateString() string
}

// BasePageTemplate - Base class for page templates.
type BasePageTemplate struct {
	titleTemplate TitleTemplate
}

func (b *BasePageTemplate) GetTemplateString() string {
	return ""
}

// TwigPageTemplate - Concrete Product for Twig page templates.
type TwigPageTemplate struct {
	BasePageTemplate
}

func (p *TwigPageTemplate) GetTemplateString() string {
	renderedTitle := p.titleTemplate.GetTemplateString()
	return fmt.Sprintf(`
		<div class="page">
			%s
			<article class="content">{{ content }}</article>
		</div>
	`, renderedTitle)
}

// PHPTemplatePageTemplate - Concrete Product for PHP page templates.
type PHPTemplatePageTemplate struct {
	BasePageTemplate
}

func (p *PHPTemplatePageTemplate) GetTemplateString() string {
	renderedTitle := p.titleTemplate.GetTemplateString()
	return fmt.Sprintf(`
		<div class="page">
			%s
			<article class="content"><?= $content; ?></article>
		</div>
	`, renderedTitle)
}

// TemplateRenderer - The Abstract Renderer interface for templates.
type TemplateRenderer interface {
	Render(templateString string, arguments map[string]interface{}) string
}

// TwigRenderer - Renderer for Twig templates.
type TwigRenderer struct{}

func (r *TwigRenderer) Render(templateString string, arguments map[string]interface{}) string {
	for key, value := range arguments {
		placeholder := fmt.Sprintf("{{ %s }}", key)
		templateString = strings.ReplaceAll(templateString, placeholder, fmt.Sprintf("%v", value))
	}
	return templateString
}

// PHPTemplateRenderer - Renderer for PHP templates.
type PHPTemplateRenderer struct{}

func (r *PHPTemplateRenderer) Render(templateString string, arguments map[string]interface{}) string {
	for key, value := range arguments {
		placeholder := fmt.Sprintf("<?= $%s; ?>", key)
		templateString = strings.ReplaceAll(templateString, placeholder, fmt.Sprintf("%v", value))
	}
	return templateString
}

// Page - Client code to render the page with templates.
type Page struct {
	title   string
	content string
}

func NewPage(title, content string) *Page {
	return &Page{title: title, content: content}
}

func (p *Page) Render(factory TemplateFactory) string {
	pageTemplate := factory.CreatePageTemplate()
	renderer := factory.GetRenderer()
	return renderer.Render(pageTemplate.GetTemplateString(), map[string]interface{}{
		"title":   p.title,
		"content": p.content,
	})
}

func main() {
	page := NewPage("Sample page", "This is the body.")

	fmt.Println("Testing actual rendering with the PHPTemplate factory:")
	fmt.Println(page.Render(&PHPTemplateFactory{}))
	fmt.Println()

}

File: patterns/Creational/AbstractFactory/JavaScript/RealWorldExample/WebTemplate.js
// TemplateFactory.js - Abstract Factory for creating templates
class TemplateFactory {
    createTitleTemplate() {
      throw new Error("Method 'createTitleTemplate()' must be implemented.");
    }
  
    createPageTemplate() {
      throw new Error("Method 'createPageTemplate()' must be implemented.");
    }
  
    getRenderer() {
      throw new Error("Method 'getRenderer()' must be implemented.");
    }
  }
  
  // TwigTemplateFactory.js - Concrete Factory for Twig templates
  class TwigTemplateFactory extends TemplateFactory {
    createTitleTemplate() {
      return new TwigTitleTemplate();
    }
  
    createPageTemplate() {
      return new TwigPageTemplate(this.createTitleTemplate());
    }
  
    getRenderer() {
      return new TwigRenderer();
    }
  }
  
  // PHPTemplateFactory.js - Concrete Factory for PHP templates
  class PHPTemplateFactory extends TemplateFactory {
    createTitleTemplate() {
      return new PHPTemplateTitleTemplate();
    }
  
    createPageTemplate() {
      return new PHPTemplatePageTemplate(this.createTitleTemplate());
    }
  
    getRenderer() {
      return new PHPTemplateRenderer();
    }
  }
  
  // TitleTemplate.js - Abstract product interface for title templates
  class TitleTemplate {
    getTemplateString() {
      throw new Error("Method 'getTemplateString()' must be implemented.");
    }
  }
  
  // TwigTitleTemplate.js - Concrete product for Twig title templates
  class TwigTitleTemplate extends TitleTemplate {
    getTemplateString() {
      return "<h1>{{ title }}</h1>";
    }
  }
  
  // PHPTemplateTitleTemplate.js - Concrete product for PHP title templates
  class PHPTemplateTitleTemplate extends TitleTemplate {
    getTemplateString() {
      return "<h1><?= \$title; ?></h1>";
    }
  }
  
  // PageTemplate.js - Abstract product interface for page templates
  class PageTemplate {
    getTemplateString() {
      throw new Error("Method 'getTemplateString()' must be implemented.");
    }
  }
  
  // BasePageTemplate.js - Base class for page templates
  class BasePageTemplate extends PageTemplate {
    constructor(titleTemplate) {
      super();
      this.titleTemplate = titleTemplate;
    }
  }
  
  // TwigPageTemplate.js - Concrete product for Twig page templates
  class TwigPageTemplate extends BasePageTemplate {
    getTemplateString() {
      const renderedTitle = this.titleTemplate.getTemplateString();
      return `
        <div class="page">
          ${renderedTitle}
          <article class="content">{{ content }}</article>
        </div>
      `;
    }
  }
  
  // PHPTemplatePageTemplate.js - Concrete product for PHP page templates
  class PHPTemplatePageTemplate extends BasePageTemplate {
    getTemplateString() {
      const renderedTitle = this.titleTemplate.getTemplateString();
      return `
        <div class="page">
          ${renderedTitle}
          <article class="content"><?= \$content; ?></article>
        </div>
      `;
    }
  }
  
  // TemplateRenderer.js - Abstract renderer for templates
  class TemplateRenderer {
    render(templateString, params = {}) {
      throw new Error("Method 'render()' must be implemented.");
    }
  }
  
  // TwigRenderer.js - Renderer for Twig templates
  class TwigRenderer extends TemplateRenderer {
    render(templateString, params = {}) {
      // Replacing {{ variable }} placeholders with actual arguments
      return templateString.replace(/\{\{ (.*?) \}\}/g, (_, key) => {
        return key in params ? params[key] : '';
      });
    }
  }
  
  // PHPTemplateRenderer.js - Renderer for PHP templates
  class PHPTemplateRenderer extends TemplateRenderer {
    render(templateString, params = {}) {
      let result = templateString;
      for (const [key, value] of Object.entries(params)) {
        result = result.replace(new RegExp(`\\\$${key}`, 'g'), value);
      }
      return result;
    }
  }
  
  
  // Page.js - Client code to render the page with templates
  class Page {
    constructor(title, content) {
      this.title = title;
      this.content = content;
    }
  
    render(factory) {
      const pageTemplate = factory.createPageTemplate();
      const renderer = factory.getRenderer();
      return renderer.render(pageTemplate.getTemplateString(), {
        title: this.title,
        content: this.content,
      });
    }
  }
  
  // Usage
  const page = new Page("Sample Page", "This is the body.");
  
  console.log("Testing rendering with the PHPTemplate factory:");
  console.log(page.render(new PHPTemplateFactory()));
  console.log("\n");
  